### PR TITLE
[Enhancement kbss-cvut/termit-ui#482] Support sending invitation links to new users

### PR DESCRIPTION
**Pull request by lukaskabc**
Implements kbss-cvut/termit-ui#482

- PasswordChangeService merged with UserService to avoid dependency loop  
- AdminBasedRegistrationController enabled for all profiles, it is now specific for any user registration by admin

### CODE REVIEW

File: src/main/java/cz/cvut/kbss/termit/rest/AdminBasedRegistrationController.java [hunk 1]
```diff
@@ -56,11 +60,17 @@ public AdminBasedRegistrationController(UserService userService) {
...
 
+    @Operation(security = {@SecurityRequirement(name="bearer-key")},
+               description = "Creates a new user account. If the password is blank, the account is locked, and an email will be sent to the new user with a link to create a password.")
+    @ApiResponses({
+            @ApiResponse(responseCode = "201", description = "User created"),
+            @ApiResponse(responseCode = "409", description = "User data are invalid")
+    })
     @PreAuthorize("hasRole('" + SecurityConstants.ROLE_ADMIN + "')")
-    @PostMapping(consumes = {MediaType.APPLICATION_JSON_VALUE, JsonLd.MEDIA_TYPE})
+    @PutMapping(consumes = {MediaType.APPLICATION_JSON_VALUE, JsonLd.MEDIA_TYPE})
```
Comment (ledsoft) [inline src/main/java/cz/cvut/kbss/termit/rest/AdminBasedRegistrationController.java]: Why did you switch from POST to PUT for creating an account?

File: src/main/java/cz/cvut/kbss/termit/service/business/UserService.java [hunk 2]
```diff
@@ -215,7 +258,7 @@ public void unlock(UserAccount account, String newPassword) {
     }
 
     private void ensureNotOwnAccount(UserAccount account, String operation) {
-        if (securityUtils.getCurrentUser().equals(account)) {
+        if (securityUtils.isAuthenticated() && securityUtils.getCurrentUser().equals(account)) {
```
Comment (ledsoft) [inline src/main/java/cz/cvut/kbss/termit/service/business/UserService.java:260]: Why did you add the first part of the check? The second part should be enough for the validation.

File: src/main/java/cz/cvut/kbss/termit/service/business/UserService.java [hunk 3]
```diff
@@ -307,4 +350,47 @@ public List<RdfsResource> getManagedAssets(@NonNull UserAccount user) {
...
+    @Transactional
+    public void requestPasswordReset(String username) {
+        final UserAccount account = repositoryService.findByUsername(username)
+                                               .orElseThrow(() -> NotFoundException.create(UserAccount.class, username));
+        PasswordChangeRequest request = createPasswordChangeRequest(account);
+        passwordChangeNotifier.sendPasswordResetEmail(request);
+    }
+
+    private boolean isValid(PasswordChangeRequest request) {
+        return request.getCreatedAt().plus(securityConfig.getPasswordChangeRequestValidity()).isAfter(Instant.now());
```
Comment (ledsoft) [inline src/main/java/cz/cvut/kbss/termit/service/business/UserService.java]: I suggest using `Utils.timestamp` which truncates the `Instant.now` to millis for easier processing (it would be good to use it for creating PasswordChangeRequests as well).

**Review by ledsoft — COMMENTED** [review]
See the inline comments

File: src/main/java/cz/cvut/kbss/termit/rest/AdminBasedRegistrationController.java (hunk 1 above)
Comment (lukaskabc) [reply src/main/java/cz/cvut/kbss/termit/rest/AdminBasedRegistrationController.java]: Currently, `AdminBasedRegistrationController` is mapped to the same path as [`FreeRegistrationController`](https://github.com/kbss-cvut/termit/blob/da05402700f6ad7e7937e2bd46cd12657fe8c180/src/main/java/cz/cvut/kbss/termit/rest/FreeRegistrationController.java#L50) and `FreeRegistrationController.createUser` shared the same mapping with `AdminBasedRegistrationController.createUser`, wasn't a problem because only one was always active (based on profile `admin-registration-only`)  
But since I set `AdminBasedRegistrationController` as always active, I had to remove the mapping conflict.  
Another option is to change the mapping path to something like `/admin/users`.

File: src/main/java/cz/cvut/kbss/termit/service/business/UserService.java (hunk 2 above)
Comment (lukaskabc) [reply src/main/java/cz/cvut/kbss/termit/service/business/UserService.java:260]: So that I can reuse the logic to unlock the user account when resetting the password.
There is no authenticated user at that time.

File: src/main/java/cz/cvut/kbss/termit/rest/AdminBasedRegistrationController.java (hunk 1 above)
Comment (ledsoft) [reply src/main/java/cz/cvut/kbss/termit/rest/AdminBasedRegistrationController.java]: Ok, in that case I vote to change the path mapping to `/admin/users`

File: src/main/java/cz/cvut/kbss/termit/service/business/UserService.java (hunk 2 above)
Comment (ledsoft) [reply src/main/java/cz/cvut/kbss/termit/service/business/UserService.java:260]: I don't understand. `securityUtils.getCurrentUser()` will not return user if they are not authenticated anyway, so the `securityUtils.isAuthenticated` seems unnecessary.

Comment (ledsoft) [reply src/main/java/cz/cvut/kbss/termit/service/business/UserService.java:260]: Or is it to prevent an exception from being thrown?

Comment (lukaskabc) [reply src/main/java/cz/cvut/kbss/termit/service/business/UserService.java:260]: > Or is it to prevent an exception from being thrown?

NullPointerException, yes  
or I could change the equation to `account.equals(securityUtils.getCurrentUser())`

Comment (ledsoft) [reply src/main/java/cz/cvut/kbss/termit/service/business/UserService.java:260]: Would that help, though? If you look at the `SecurityUtils.getCurrentUser` code, the NPX would probably be thrown there (or maybe ClassCastException). Anyway, I now see the reasoning. Let's keep this as is.

**Review by ledsoft — APPROVED** [review]
LGTM
